File: scripts/l10n-export.mjs
import { execFileSync } from "node:child_process";
import { cp, mkdir, mkdtemp, readdir, rm } from "node:fs/promises";
import { basename, dirname, extname, join, relative, resolve } from "node:path";

const sourceRoot = resolve("src");
const outputRoot = resolve("l10n");
const temporaryRoot = await mkdtemp(join(resolve("."), ".fleximark-l10n-"));
const temporarySource = join(temporaryRoot, "src");

async function copySources(directory) {
  for (const entry of await readdir(directory, { withFileTypes: true })) {
    const source = join(directory, entry.name);
    if (entry.isDirectory()) {
      await copySources(source);
      continue;
    }
    if (![".mts", ".cts", ".ts", ".js"].includes(extname(entry.name))) {
      continue;
    }

    const relativePath = relative(sourceRoot, source);
    const targetDirectory = join(temporarySource, dirname(relativePath));
    const targetName = basename(relativePath).replace(/\.(?:mts|cts)$/, ".ts");
    await mkdir(targetDirectory, { recursive: true });
    await cp(source, join(targetDirectory, targetName));
  }
}

try {
  await copySources(sourceRoot);
  execFileSync(
    process.execPath,
    [
      resolve("node_modules/@vscode/l10n-dev/dist/cli.js"),
      "export",
      "--outDir",
      outputRoot,
      temporarySource,
    ],
    { stdio: "inherit" },
  );
} finally {
  await rm(temporaryRoot, { recursive: true, force: true });
}

File: scripts/l10n_export.py
from __future__ import annotations

import shutil
import tempfile
from pathlib import Path

from _tools import ROOT, run, script_entrypoint


SOURCE_EXTENSIONS = {".mts", ".cts", ".ts", ".js"}


def export_localization() -> None:
    source_root = ROOT / "adapters" / "vscode" / "src"
    output_root = ROOT / "l10n"
    with tempfile.TemporaryDirectory(prefix=".fleximark-l10n-", dir=ROOT) as temp:
        temporary_source = Path(temp) / "src"
        for source in source_root.rglob("*"):
            if not source.is_file() or source.suffix not in SOURCE_EXTENSIONS:
                continue
            relative = source.relative_to(source_root)
            target_name = (
                relative.with_suffix(".ts")
                if relative.suffix in {".mts", ".cts"}
                else relative
            )
            target = temporary_source / target_name
            target.parent.mkdir(parents=True, exist_ok=True)
            shutil.copy2(source, target)

        run(
            "node",
            "node_modules/@vscode/l10n-dev/dist/cli.js",
            "export",
            "--outDir",
            str(output_root),
            str(temporary_source),
        )

    bundle = output_root / "bundle.l10n.json"
    contents = bundle.read_text(encoding="utf-8")
    if not contents.endswith("\n"):
        contents += "\n"
    bundle.write_bytes(contents.encode("utf-8"))


if __name__ == "__main__":
    script_entrypoint(export_localization)
